File: account/src/main/java/serp/project/account/ui/controller/SubscriptionController.java
/**
 * Author: QuanTuanHuy
 * Description: Part of Serp Project
 */

package serp.project.account.ui.controller;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import serp.project.account.core.domain.dto.request.*;
import serp.project.account.core.usecase.SubscriptionUseCase;
import serp.project.account.kernel.utils.AuthUtils;

@RequiredArgsConstructor
@RestController
@RequestMapping("/api/v1/subscriptions")
@Slf4j
public class SubscriptionController {

    private final SubscriptionUseCase organizationSubscriptionUseCase;
    private final AuthUtils authUtils;

    @PostMapping("/subscribe")
    public ResponseEntity<?> subscribe(
            @Valid @RequestBody SubscribeRequest request) {
        Long requestedBy = authUtils.getCurrentUserId().orElse(null);
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);

        log.info("POST /api/v1/subscriptions/subscribe - Subscribing to plan {}",
                tenantId, request.getPlanId());
        var response = organizationSubscriptionUseCase.subscribe(tenantId, request, requestedBy);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PostMapping("/subscribe-custom-plan")
    public ResponseEntity<?> subscribeCustomPlan(
            @Valid @RequestBody SubscribeCustomPlanRequest request) {
        Long requestedBy = authUtils.getCurrentUserId().orElse(null);
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);

        log.info("POST /api/v1/subscriptions/subscribe-custom-plan - Subscribing to custom plan {}",
                tenantId);
        var response = organizationSubscriptionUseCase.subscribeCustomPlan(tenantId, request, requestedBy);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PostMapping("/request-more-modules")
    public ResponseEntity<?> requestMoreModules(
            @Valid @RequestBody RequestMoreModulesRequest request) {
        Long requestedBy = authUtils.getCurrentUserId().orElse(null);
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);

        log.info("POST /api/v1/subscriptions/request-more-modules - Requesting more modules for organization {}",
                tenantId);
        var response = organizationSubscriptionUseCase.requestMoreModulesForPlan(tenantId, request, requestedBy);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PostMapping("/trial")
    public ResponseEntity<?> startTrial(
            @RequestParam Long planId) {
        Long requestedBy = authUtils.getCurrentUserId().orElse(null);
        Long organizationId = authUtils.getCurrentTenantId().orElse(null);

        log.info("POST /api/v1/subscriptions/trial - Starting trial for plan {}",
                organizationId, planId);
        var response = organizationSubscriptionUseCase.startTrial(organizationId, planId, requestedBy);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PutMapping("/upgrade")
    public ResponseEntity<?> upgradeSubscription(
            @Valid @RequestBody UpgradeSubscriptionRequest request) {
        Long requestedBy = authUtils.getCurrentUserId().orElse(null);
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);

        log.info("PUT /api/v1/subscriptions/upgrade - Upgrading to plan {}",
                request.getNewPlanId());
        var response = organizationSubscriptionUseCase.upgradeSubscription(tenantId, request, requestedBy);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PutMapping("/downgrade")
    public ResponseEntity<?> downgradeSubscription(
            @Valid @RequestBody DowngradeSubscriptionRequest request) {
        Long requestedBy = authUtils.getCurrentUserId().orElse(null);
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);

        log.info("PUT /api/v1/subscriptions/downgrade - Downgrading to plan {}",
                request.getNewPlanId());
        var response = organizationSubscriptionUseCase.downgradeSubscription(tenantId, request, requestedBy);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PutMapping("/cancel")
    public ResponseEntity<?> cancelSubscription(
            @Valid @RequestBody CancelSubscriptionRequest request) {
        Long requestedBy = authUtils.getCurrentUserId().orElse(null);
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);

        log.info("PUT /api/v1/subscriptions/cancel - Cancelling subscription");
        var response = organizationSubscriptionUseCase.cancelSubscription(tenantId, request, requestedBy);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PutMapping("/renew")
    public ResponseEntity<?> renewSubscription() {
        Long requestedBy = authUtils.getCurrentUserId().orElse(null);
        Long tenantId = authUtils.getCurrentTenantId().orElse(null);

        log.info("PUT /api/v1/subscriptions/renew - Renewing subscription");
        var response = organizationSubscriptionUseCase.renewSubscription(tenantId, requestedBy);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PutMapping("/{subscriptionId}/extend-trial")
    public ResponseEntity<?> extendTrial(
            @PathVariable Long subscriptionId,
            @Valid @RequestBody ExtendTrialRequest request) {
        Long requestedBy = authUtils.getCurrentUserId().orElse(null);

        log.info("PUT /api/v1/subscriptions/{}/extend-trial - Extending trial", subscriptionId);
        var response = organizationSubscriptionUseCase.extendTrial(subscriptionId, request, requestedBy);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping("/me/active")
    public ResponseEntity<?> getActiveSubscription() {
        Long organizationId = authUtils.getCurrentTenantId().orElse(null);

        log.info("GET /api/v1/subscriptions/me/active - Fetching active subscription for organization {}",
                organizationId);
        var response = organizationSubscriptionUseCase.getActiveSubscription(organizationId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping("/{subscriptionId}")
    public ResponseEntity<?> getSubscriptionById(@PathVariable Long subscriptionId) {
        log.info("GET /api/v1/subscriptions/{} - Fetching subscription", subscriptionId);
        var response = organizationSubscriptionUseCase.getSubscriptionById(subscriptionId);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping("/me/history")
    public ResponseEntity<?> getSubscriptionHistory() {
        Long organizationId = authUtils.getCurrentTenantId().orElse(null);

        log.info("GET /api/v1/subscriptions/me/history - Fetching subscription history for organization {}",
                organizationId);
        var response = organizationSubscriptionUseCase.getSubscriptionHistory(organizationId);
        return ResponseEntity.status(response.getCode()).body(response);
    }
}

File: api_gateway/src/ui/controller/account/subscription_controller.go
/*
Author: QuanTuanHuy
Description: Part of Serp Project
*/

package controller

import (
	"github.com/gin-gonic/gin"
	"github.com/serp/api-gateway/src/core/domain/constant"
	request "github.com/serp/api-gateway/src/core/domain/dto/request/account"
	service "github.com/serp/api-gateway/src/core/service/account"
	"github.com/serp/api-gateway/src/kernel/utils"
)

type SubscriptionController struct {
	subscriptionService service.ISubscriptionService
}

func (s *SubscriptionController) GetAllSubscriptions(c *gin.Context) {
	params := request.GetSubscriptionParams{
		Page:           utils.ParseIntQuery(c, "page"),
		PageSize:       utils.ParseIntQuery(c, "pageSize"),
		SortBy:         utils.ParseStringQuery(c, "sortBy"),
		SortDir:        utils.ParseStringQuery(c, "sortDir"),
		OrganizationID: utils.ParseInt64Query(c, "organizationId"),
		Status:         utils.ParseStringQuery(c, "status"),
		BillingCycle:   utils.ParseStringQuery(c, "billingCycle"),
	}

	res, err := s.subscriptionService.GetAllSubscriptions(c.Request.Context(), &params)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (s *SubscriptionController) Subscribe(c *gin.Context) {
	var req request.SubscribeRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := s.subscriptionService.Subscribe(c.Request.Context(), &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (s *SubscriptionController) SubscribeCustomPlan(c *gin.Context) {
	var req request.SubscribeCustomPlanRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := s.subscriptionService.SubscribeCustomPlan(c.Request.Context(), &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (s *SubscriptionController) RequestMoreModules(c *gin.Context) {
	var req request.RequestMoreModulesRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := s.subscriptionService.RequestMoreModules(c.Request.Context(), &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (s *SubscriptionController) StartTrial(c *gin.Context) {
	planId, ok := utils.ValidateAndParseQueryID(c, "planId")
	if !ok {
		return
	}

	res, err := s.subscriptionService.StartTrial(c.Request.Context(), planId)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (s *SubscriptionController) ActivateSubscription(c *gin.Context) {
	subscriptionId, ok := utils.ValidateAndParseID(c, "subscriptionId")
	if !ok {
		return
	}

	res, err := s.subscriptionService.ActivateSubscription(c.Request.Context(), subscriptionId)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (s *SubscriptionController) RejectSubscription(c *gin.Context) {
	subscriptionId, ok := utils.ValidateAndParseID(c, "subscriptionId")
	if !ok {
		return
	}

	var req request.RejectSubscriptionRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := s.subscriptionService.RejectSubscription(c.Request.Context(), subscriptionId, &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (s *SubscriptionController) UpgradeSubscription(c *gin.Context) {
	var req request.UpgradeSubscriptionRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := s.subscriptionService.UpgradeSubscription(c.Request.Context(), &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (s *SubscriptionController) DowngradeSubscription(c *gin.Context) {
	var req request.DowngradeSubscriptionRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := s.subscriptionService.DowngradeSubscription(c.Request.Context(), &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (s *SubscriptionController) CancelSubscription(c *gin.Context) {
	var req request.CancelSubscriptionRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := s.subscriptionService.CancelSubscription(c.Request.Context(), &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (s *SubscriptionController) RenewSubscription(c *gin.Context) {
	res, err := s.subscriptionService.RenewSubscription(c.Request.Context())
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (s *SubscriptionController) ExtendTrial(c *gin.Context) {
	subscriptionId, ok := utils.ValidateAndParseID(c, "subscriptionId")
	if !ok {
		return
	}

	var req request.ExtendTrialRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := s.subscriptionService.ExtendTrial(c.Request.Context(), subscriptionId, &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (s *SubscriptionController) ExpireSubscription(c *gin.Context) {
	subscriptionId, ok := utils.ValidateAndParseID(c, "subscriptionId")
	if !ok {
		return
	}

	res, err := s.subscriptionService.ExpireSubscription(c.Request.Context(), subscriptionId)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (s *SubscriptionController) GetActiveSubscription(c *gin.Context) {
	res, err := s.subscriptionService.GetActiveSubscription(c.Request.Context())
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (s *SubscriptionController) GetSubscriptionById(c *gin.Context) {
	subscriptionId, ok := utils.ValidateAndParseID(c, "subscriptionId")
	if !ok {
		return
	}

	res, err := s.subscriptionService.GetSubscriptionById(c.Request.Context(), subscriptionId)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (s *SubscriptionController) GetSubscriptionHistory(c *gin.Context) {
	res, err := s.subscriptionService.GetSubscriptionHistory(c.Request.Context())
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func NewSubscriptionController(subscriptionService service.ISubscriptionService) *SubscriptionController {
	return &SubscriptionController{
		subscriptionService: subscriptionService,
	}
}
